src/gennylib/include/gennylib/PhaseLoop.hpp

```diff
@@ -77,10 +78,44 @@ class IterationCompletionCheck final {
         }
     }
 
-    explicit IterationCompletionCheck(PhaseContext& phaseContext)
-        : IterationCompletionCheck(phaseContext.get<TimeSpec, false>("Duration"),
-                                   phaseContext.get<UIntSpec, false>("Repeat"),
-                                   phaseContext.isNop()) {}
+    explicit IterationChecker(PhaseContext& phaseContext)
+        : IterationChecker(phaseContext.get<TimeSpec, false>("Duration"),
+                           phaseContext.get<IntegerSpec, false>("Repeat"),
+                           phaseContext.isNop()) {
+        auto rateSpec = phaseContext.get<RateSpec, false>("Rate");
+        const auto rateLimiterName =
+            phaseContext.get<std::string, false>("RateLimiterName").value_or("defaultRateLimiter");
+
+        if (rateSpec) {
+            _rateLimiter =
+                phaseContext.workload().getRateLimiter(rateLimiterName, rateSpec.value());
+        }
+    }
+
+    bool shouldLimitRate(int64_t currentIteration) const {
+        // Only rate limit if the current iteration is a muliple of the burst size.
+        return _rateLimiter && (currentIteration % _rateLimiter->getBurstSize() == 0);
+    }
+
+    void limitRate(const int64_t currentIteration, const Orchestrator& o, const PhaseNumber inPhase) const {
+        // This function is called after each iteration, so we never rate limit the
+        // first iteration. This means the number of completed operations is always
+        // `n * GlobalRateLimiter::_burstSize + m` instead of an exact multiple of
+        // _burstSize. `m` here is the number of threads using the rate limiter.
+        if (shouldLimitRate(currentIteration)) {
+            while (true) {
+                auto success = _rateLimiter->consume();;
+                if (!success && (o.currentPhase() == inPhase)) {
+                    // Sleep for a little bit. This is not efficient if there's a large number of
+                    // threads. If this is a problem, we can use exponential back off or wait
+                    // based on the number of actors using this rate limiter.
+                    std::this_thread::sleep_for(std::chrono::nanoseconds(_rateLimiter->getRate()));
```

Comment:
thinking a bit more about this....so this works well for bursty workloads, but will perf be good enough here to support a fast but constant-rate workload? E.g. imagine wanting to do `"1 per 100 microseconds"`. If we have 10k threads all competing for the same token will we actually be able to achieve that rate given the sleeps? May warrant a bit of playing around with it and/or a benchmark test.

Reply:
I believe yes. With the current approach, we only need one thread, any thread, to get waken up per 100us for the workload to proceed. The scheduler should be smart enough to do this. Additionally, since sleep() and the rate limiter are both lock free, the number of threads is irrelevant, 100k threads should behave similarly to 10k threads.

With the condition variable approach, you'd have 10k threads all trying to acquire some mutex. Acquiring a mutex takes a non-trivial amount of time (50ns according to this guy: https://stackoverflow.com/a/49712993), so in the pathological worst case if the threads are all scheduled sequentially when one thread is in the critical section, you'd need 50ns * 10k = 500us for the thread doing the workload to proceed.

I'll write up a benchmark to test out that the rate limiter is able to proceed with >10k threads and a 1 per 100 microseconds rate.

For the BigUpdate workload, I saw no performance difference between this approach and the previous ad-hoc approach: https://evergreen.mongodb.com/version/5c4a3b1a3066153a37fb754a (see the first execution). But funnily, I believe the BigUpdate workload currently isn't using the rate limiter at all, because the number of actual operations is lower than what the rate limiter requires.

I've restarted the task with a lower rate limit to verify that the rate limiter does anything at all.

Reply:
> I'll write up a benchmark to test out that the rate limiter is able to proceed with >10k threads and a 1 per 100 microseconds rate.

Thanks - if the benchmark bares it out no need to speculate further. 

Thinking aloud: A benchmark test could have 10k threads that increment an atomic counter rate-limited to "1 per 100 microseconds" and runs for 5000000 microseconds (5 seconds). Counter should be exactly 50000 at the end.